src/node.rs

```diff
@@ -346,6 +236,85 @@ impl<K: ArtKey, V: Default, const MAX_PARTIAL_LEN: usize> ArtNode<K, V, MAX_PART
+        new_node_header.partial.data[..max_copy_len]
+            .copy_from_slice(&old_node_header.partial.data[..max_copy_len]);
+        new_node_header.partial.len = prefix_mismatch_pos as u32;
+
+        // Note: The match may exceed the maximum of the vector store, so take the minimum of both.
+        // Artful uses mixed compression, So the actual storage length of partial can exceed the
+        // maximum value of the vector.
+        // pessimistic compression
+        if old_node_header.partial.len as usize <= MAX_PARTIAL_LEN {
+            let old_node_header = old_node.header_mut();
+            let old_node_byte = old_node_header.partial.data[prefix_mismatch_pos];
+            old_node_header.partial.len -= (prefix_mismatch_pos + 1) as u32;
+            old_node_header
+                .partial
+                .data
+                .rotate_left(prefix_mismatch_pos + 1);
+            // unsafe {
+            //     std::ptr::copy(
+            //         old_node_header
+            //             .partial
+            //             .data
+            //             .as_ptr()
+            //             .offset(prefix_mismatch_pos as isize + 1),
+            //         old_node_header.partial.data.as_mut_ptr(),
+            //         min(MAX_PARTIAL_LEN, old_node_header.partial.len as usize),
+            //     );
+            // }
+            self.insert_child((old_node_byte, true), old_node);
+            self.insert_child(
+                ArtKeyVerifier::valid(key.get_bytes(), depth + prefix_mismatch_pos),
+                ArtNode::leaf(key, val),
+            );
+
+            return;
+        }
+
+        // optimistic compression
+        // TODO: optimization the Header::default() to zero size.
+        let mut old_node_header = std::mem::take(old_node.header_mut());
+        let leaf =
+            ArtNode::minimum_child(&old_node).expect("the inner node get minimum child fail");
+
+        let leaf_key_bytes = leaf.key.get_bytes();
+        let valid_key = ArtKeyVerifier::valid(leaf_key_bytes, depth + prefix_mismatch_pos);
+
+        // TODO add proof.
+        old_node_header.partial.len -= (prefix_mismatch_pos + 1) as u32;
+        let max_copy_len = min(MAX_PARTIAL_LEN, old_node_header.partial.len as usize);
+        let start = depth + prefix_mismatch_pos + 1;
+        let end = start + max_copy_len;
+        old_node_header.partial.data[..max_copy_len].copy_from_slice(&leaf_key_bytes[start..end]);
+
+        std::mem::swap(&mut old_node_header, old_node.header_mut());
+        // unsafe {
+        //     copy_nonoverlapping(
+        //         key_bytes
+        //             .as_ptr()
+        //             .offset((depth + prefix_mismatch_pos + 1) as isize),
+        //         old_node.header_mut().partial.data.as_mut_ptr(),
+        //         min(MAX_PARTIAL_LEN, old_node.header().partial.len as usize),
+        //     )
+        // }
```

Comment:
`minimum_child ` needs to borrow from `old_node`, but then relies on the return value of `minimum_child` to set the header of 'old_node', which is not allowed by Rust's memory model.

The idea here is to first remove the ownership of the `old_node ` header using `std::mem::take`, then do the following, and finally swap through `std ::mem::swap`. This works, but results in a memory allocation for the header, since `std::mem::take` executes the `header :default` once.